rust/core-lib/src/editor.rs

```diff
@@ -183,8 +185,40 @@ impl<W: Write + Send + 'static> Editor<W> {
     }
 
     fn insert(&mut self, s: &str) {
-        let sel_interval = Interval::new_closed_open(self.view.sel_min(), self.view.sel_max());
-        let new_cursor = self.view.sel_min() + s.len();
+        let start = self.view.sel_min();
+        let mut end = self.view.sel_max();
+        let new_cursor = start + s.len();
+
+        // process soft spans in front of the cursor
+        let iv_ahead = Interval::new_closed_open(start, self.soft_spans.len());
+        let soft_spans = self.soft_spans.subseq(iv_ahead);
+        for (interval, _) in soft_spans.iter() {
+            // only interested in soft spans that are directly in front of the cursor
+            if interval.start() != 0 {
+                break
+            }
+
+            let mut cursor = Cursor::new(&self.text, start);
+            let are_eql = compare_cursor_str(&mut cursor, s);
+
+            // are_eql is checked for 1 codepoint long inserts and cursor position is checked
+            // to also match longer inserts partially
```

Comment:
I have noticed that the codebase has only few inline comments. If this is preferred I am also fine with removing them